src_c/display.c

```diff
@@ -2021,14 +2066,140 @@ static PyObject *
-    else
-        result = SDL_SetWindowFullscreen(win, 0);
+
+    flags = SDL_GetWindowFlags(win) &
+            (SDL_WINDOW_FULLSCREEN | SDL_WINDOW_FULLSCREEN_DESKTOP);
+
+    SDL_VERSION(&info.version);
+    SDL_GetWindowWMInfo(win, &info);
+    switch(info.subsystem) {
+
+      // if we get this to work correctly with more systems, move them here
+      case SDL_SYSWM_X11:
+	break; 
+
+      case SDL_SYSWM_WINDOWS:
+#if SDL_VERSION_ATLEAST(2, 0, 3)
+      case SDL_SYSWM_WINRT:
+#endif
+      case SDL_SYSWM_DIRECTFB:
+      case SDL_SYSWM_COCOA:
+      case SDL_SYSWM_UIKIT:
+#if SDL_VERSION_ATLEAST(2, 0, 2)
+      case SDL_SYSWM_WAYLAND:
+      case SDL_SYSWM_MIR:
+#endif
+#if SDL_VERSION_ATLEAST(2, 0, 4)
+      case SDL_SYSWM_ANDROID:
+#endif
+#if SDL_VERSION_ATLEAST(2, 0, 5)
+      case SDL_SYSWM_VIVANTE:
+#endif
+    // everything that doesn't work will fallthrough to here
+    // this includes windows and mac as of now.
+    // not raising an exception here is deliberate,
+    // for backwards compatibility
+    case SDL_SYSWM_UNKNOWN:
+	return PyInt_FromLong(-1);
+    }
+
+    display_surface = pg_GetDefaultWindowSurface();
+
+    // could also take the size of the old display surface
+    SDL_GetWindowSize(win, &window_w, &window_h);
+    window_display = SDL_GetWindowDisplayIndex(win);
+    if (SDL_GetDesktopDisplayMode(window_display, &display_mode) != 0) {
+        return RAISE(pgExc_SDLError, SDL_GetError());
+    }
+
+    /*
+      if (pg_renderer != NULL) {
+        SDL_RenderGetLogicalSize(pg_renderer, &w, &h);
+    }
+    */
+    if (state->using_gl) {
+        p_glViewport = (GL_glViewport_Func)SDL_GL_GetProcAddress("glViewport");
+        SDL_GL_GetDrawableSize(win, &w, &h);
+    } else {
+        w = display_surface->surf->w;
+        h = display_surface->surf->h;
+    }
+
+    if (flags & SDL_WINDOW_FULLSCREEN) {
+        /* TOGGLE FULLSCREEN OFF */
+
+        if (pg_renderer != NULL) {
+            int scale = 1;
+            int xscale, yscale;
+
+            xscale = window_w / w;
+            yscale = window_h / h;
+            scale = xscale < yscale ? xscale : yscale;
+            if (scale < 1)
+                scale = 1;
+
+            result = SDL_SetWindowFullscreen(win, 0);
+            SDL_SetWindowSize(win, w * scale, h * scale);
+            SDL_RenderSetLogicalSize(pg_renderer, w, h);
+        } else if (state -> using_gl) {
+            result = SDL_SetWindowFullscreen(win, 0);
+            SDL_GL_MakeCurrent(win, state-> gl_context);
+            if (state->toggle_windowed_w>0
+                && state->toggle_windowed_h>0) {
+                p_glViewport(0, 0,
+                             state->toggle_windowed_w,
+                             state->toggle_windowed_h);
+                state->toggle_windowed_w = 0;
+                state->toggle_windowed_h = 0;
+            }
+        } else if (flags == SDL_WINDOW_FULLSCREEN_DESKTOP) {
```

Comment:
This should probably say `flags & SDL_WINDOW_FULLSCREEN_DESKTOP` (if SDL_WINDOW_FULLSCREEN *can* be set at the same time)

Reply:
if flags == SDL_WINDOW_FULLSCREEN
then flags & SDL_WINDOW_FULLSCREEN_DESKTOP == SDL_WINDOW_FULLSCREEN

Reply:
This should probably say `flags & SDL_WINDOW_FULLSCREEN_DESKTOP == SDL_WINDOW_FULLSCREEN_DESKTOP` in that case

Reply:
Sorry. I see now that SDL_WINDOW_FULLSCREEN_DESKTOP includes the flag SDL_WINDOW_FULLSCREEN